vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/CacheInterceptor.java

```diff
@@ -0,0 +1,275 @@
+
+        // Cannot expire if not set
+        if (!maxAge.isPresent()) {
+            return false;
+        }
+
+        // Check that this entry has not expired if age was set
+        return System.currentTimeMillis() > cacheValue.createdAt + maxAge.get();
+    }
+
+    private boolean isValidEtag(HttpRequestImpl request, CacheValue cacheValue) {
+        String requestEtag = request.headers().get("ETag");
+
+        if (requestEtag == null) {
+            return true;
+        }
+
+        String responseEtag = cacheValue.value.headers().get("ETag");
+
+        if (responseEtag == null) {
+            return true;
+        }
+
+        // If ETags are different, value should be considered expired
+        if  (!responseEtag.equals(requestEtag)) {
+            return false;
+        }
+
+        return true;
+    }
+
+    private Optional<Integer> parseMaxAge(Set<String> cacheControl) {
+        Optional<String> maxAge = cacheControl.stream().
+                filter(c -> c.startsWith("max-age")).
+                findFirst();
+
+        if (maxAge.isPresent()) {
+            String[] maxAgeParts = maxAge.get().split("=");
+            if (maxAgeParts.length > 1) {
+                try {
+                    return Optional.of(Integer.parseInt(maxAgeParts[1]));
+                }
+                catch (NumberFormatException nfe) {
+                    return Optional.empty();
+                }
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    private void handleCacheMiss(HttpContext event, CacheKey cacheKey) {
+        Handler<AsyncResult<HttpResponse<Object>>> responseHandler = event.getResponseHandler();
+        event.setResponseHandler(r -> {
+            if (r.succeeded()) {
+                HttpResponse<Object> response = r.result();
+
+                // Cache response and add it as most recently used
+                synchronized (this) {
+                    lru.add(cacheKey);
+                    cache.put(cacheKey, new CacheValue(response));
+                }
+            }
+            responseHandler.handle(r);
+        });
+        event.next();
+    }
+
+    /**
+     * Removes least recently used element from the cache
+     */
+    private void invalidate() {
+        if (lru.size() > options.getMaxEntries()) {
+            if (log.isDebugEnabled()) {
+                log.debug(String.format("Cache is full, size is %d", lru.size()));
+            }
+            synchronized (this) {
+                Iterator<CacheKey> it = lru.iterator();
+                if (it.hasNext()) {
+                    CacheKey lruKey = it.next();
+                    it.remove();
+                    cache.remove(lruKey);
+                }
+            }
+        }
+    }
+
+    public void flush() {
+        synchronized (this) {
+            lru.clear();
+            cache.clear();
+        }
+    }
+
+    private class CacheKey {
+        private final HttpMethod method;
```

Comment:
In multi-tenant applications, caching at the `WebClient` level could be a security problem.

Usually you would create a single `WebClient` for all users. Then at request time you'd set, for example, a token in HTTP headers. Now when user B needs `/private/myuserinfo` it could get data from user A because the `CacheKey` would be the same.

Or am I missing something?

Reply:
I think that in general you're right, and that's the issue @vietj was raising in terms of how "shared" this cache should be. 
Don't see how this issue could be resolved, though, since clients could use any `X-*` header for their token. The only idea I have is to provide a specific method to set the token, like `webClient.putAuthenticationToken("a", "b")`, and then we can recognise such cases.
Any other ideas?

Reply:
hi, regarding caching you should follow the way proxies work, cache keys in most cases are configurable as the concatenation of some known elements e.g.:

* scheme (GET/POST/etc)
* host
* uri
* Authorization header
* Cookie

For example NGNIX has this on their docs:

http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_cache_key

perhaps having a a way to configure this would be nice, say a list of strings?

Reply:
Possible, that's what my first implementation looked like, but a bit dangerous if user provides configuration that doesn't make much sense, like only `["method"]` or only `["url"]`

But as far as I understand having the ability to configure cache key doesn't solve privacy problem.

Or do you propose that we give users ability to configure cache key based on headers, and then move this problem to them?

Reply:
You're correct, this will not solve the privacy issue. I guess if we're just talking privacy, the bare minimum would be no cache if a response contains a cookie or was made with a cookie or authorization header

Reply:
I think we can have caching with such cookie if we give the opportunity in the API to have control over the scope

Reply:
I think the idea would be to have one WebClient per session or a similar design, I believe the crux is that we could have two level of caching : per session and shared

Reply:
yes something like that, that would manage cookies, caching, etc...

Reply:
If I get some definitions I think I can implement this, but I'm not sure if it should be part of this PR or not.